pkg/controller/daemon_controller.go

```diff
@@ -0,0 +1,432 @@
+			glog.V(2).Infof("Failed creation, decrementing expectations for controller %q/%q", dc.Namespace, dc.Name)
+			dm.expectations.CreationObserved(dcKey)
+			util.HandleError(err)
+		}
+	}
+
+	glog.Infof("podsToDelete: %+v", podsToDelete)
+	for i := range podsToDelete {
+		if err := dm.podControl.deletePod(dc.Namespace, podsToDelete[i]); err != nil {
+			glog.V(2).Infof("Failed deletion, decrementing expectations for controller %q/%q", dc.Namespace, dc.Name)
+			dm.expectations.DeletionObserved(dcKey)
+			util.HandleError(err)
+		}
+	}
+}
+
+func (dm *DaemonManager) updateDaemonStatus(dc *api.DaemonController) {
+	glog.Infof("Updating daemon status")
+	nodeToDaemonPods, err := dm.getNodesToDaemonPods(dc)
+	if err != nil {
+		glog.Errorf("Error getting node to daemon pod mapping for daemon %+v: %v", dc, err)
+	}
+
+	nodeList, err := dm.nodeStore.List()
+	if err != nil {
+		glog.Errorf("Couldn't get list of nodes when adding daemon %+v: %v", dc, err)
+	}
+
+	var desiredNumberScheduled, currentNumberScheduled, numberMisscheduled int
+	for i := range nodeList.Items {
+		nodeSelector := labels.Set(dc.Spec.Template.Spec.NodeSelector).AsSelector()
+		shouldRun := nodeSelector.Matches(labels.Set(nodeList.Items[i].Labels))
+		numDaemonPods := len(nodeToDaemonPods[nodeList.Items[i].Name])
+		if shouldRun {
+			desiredNumberScheduled++
+			if numDaemonPods == 1 {
+				currentNumberScheduled++
+			}
+		} else if numDaemonPods >= 1 {
+			numberMisscheduled++
+		}
+	}
+
+	glog.Infof("Counts: %d %d %d", desiredNumberScheduled, currentNumberScheduled, numberMisscheduled)
+	err = storeDaemonStatus(dm.kubeClient.DaemonControllers(dc.Namespace), dc, desiredNumberScheduled, currentNumberScheduled, numberMisscheduled)
+	if err != nil {
+		glog.Errorf("Error storing status for daemon %+v: %v", dc, err)
+	}
+}
+
+func (dm *DaemonManager) syncDaemonController(key string) error {
+	startTime := time.Now()
+	defer func() {
+		glog.V(4).Infof("Finished syncing daemon %q (%v)", key, time.Now().Sub(startTime))
+	}()
+	obj, exists, err := dm.dcStore.Store.GetByKey(key)
+	if !exists {
+		glog.Infof("Daemon Controller has been deleted %v", key)
+		dm.expectations.DeleteExpectations(key)
+		return nil
+	}
+	if err != nil {
+		glog.Infof("Unable to retrieve dc %v from store: %v", key, err)
+		dm.queue.Add(key)
+		return err
+	}
+	dc := obj.(*api.DaemonController)
+
+	// Don't process a daemon controller until all it's and deletions have been processed. For example
+	// if dc foo asked for 3 new daemon pods in the previous call to manageDaemons, then we do not want
+	// to call manageDaemons on foo until the daemons have been created.
+	dcKey, err := controllerKeyFunc(dc)
+	if err != nil {
+		glog.Errorf("Couldn't get key for object %+v: %v", dc, err)
+	}
+	dcNeedsSync := dm.expectations.SatisfiedExpectations(dcKey)
+	if dcNeedsSync {
+		dm.manageDaemons(dc)
+	}
+
+	dm.updateDaemonStatus(dc)
```

Comment:
Since manageDaemons and updateDaemonStatus do almost the same amount of work, I wonder if they could be combined. Just thinking out loud, not asking for a change.